### PR TITLE
Make IPackageInstallerService implementation lazy

### PR DESCRIPTION
**Pull request by sharwell**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/833267

The internal bug is a performance bug caused by `PackageInstallerService.StartWorking` eagerly initializing `PackageSources`, even in scenarios where the property value is never used. This change updates the implementation to instead load the package sources only when needed (e.g. for a code action).

### CODE REVIEW

**Comment by sharwell** [comment]
Blocked on https://github.com/dotnet/roslyn-analyzers/pull/2400, tentative target 16.3.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 1]
```diff
@@ -69,25 +72,73 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
...
+            {
+                // These exceptions can happen when the nuget.config file is broken.
+                packageSources = ImmutableArray<PackageSource>.Empty;
+            }
+
+            var previousPackageSources = ImmutableInterlocked.InterlockedCompareExchange(ref _packageSources, packageSources, default);
+            if (previousPackageSources != null)
+            {
+                // Another thread already initialized _packageSources
+                packageSources = previousPackageSources;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: 1. couldn' this all be simplified to: 

```c#
var previousPackageSources = ...;
return _packageSources;
```

?

In other words, you' basically try to assign into _packageSources.  Either you win, or someone else won.  In either case, you return the value of hte winner.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 2]
```diff
@@ -48,9 +48,12 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
...
 
-        // We refer to the package services through proxy types so that we can
-        // delay loading their DLLs until we actually need them.
-        private IPackageServicesProxy _packageServices;
+        private readonly Lazy<IVsPackageInstallerServices> _packageInstallerServices;
+        private readonly Lazy<IVsPackageInstaller2> _packageInstaller;
+        private readonly Lazy<IVsPackageUninstaller> _packageUninstaller;
+        private readonly Lazy<IVsPackageSourceProvider> _packageSourceProvider;
+
+        private ImmutableArray<PackageSource> _packageSources;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: based on hte impl below, why not just use an AsyncLazy here?  It seems clearer and better expressed the intent.  Thoughts?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 3]
```diff
@@ -69,25 +72,73 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
...
+            }
+
+            try
+            {
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
+                    .Select(r => new PackageSource(r.Key, r.Value))
+                    .ToImmutableArrayOrEmpty();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: SelectAsArray?

File: src/VisualStudio/Core/Def/SymbolSearch/VisualStudioSymbolSearchService.cs [hunk 4]
```diff
@@ -76,18 +76,9 @@ private void OnPackageSourcesChanged(object sender, EventArgs e)
...
             // Always pull down the nuget.org index.  It contains the MS reference assembly index
             // inside of it.
-            var allSources = sources.Concat(new PackageSource(
-                SymbolSearchUpdateEngine.NugetOrgSource, source: null));
-
-            foreach (var source in allSources)
-            {
-                Task.Run(() => UpdateSourceInBackgroundAsync(source.Name));
-            }
+            Task.Run(() => UpdateSourceInBackgroundAsync(SymbolSearchUpdateEngine.NugetOrgSource));
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/SymbolSearch/VisualStudioSymbolSearchService.cs]: why this change?  the intent here was to allow other parties to participate in the search.  we coded our end up to make that possible, did the other interested parties retract their interest in playing in tihs space with us?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: No. It could be set back to null by the event handler.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 2 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: I opted for optimistic concurrency control to avoid unknown interactions between initialization and vs-threading (AsyncLazy is not known to be safe in combination with an application using JTF, so I avoid it unless it provides a benefit not achievable through other means).

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: Ah, thanks.  Could we have a comment on that?

File: src/VisualStudio/Core/Def/SymbolSearch/VisualStudioSymbolSearchService.cs (hunk 4 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/SymbolSearch/VisualStudioSymbolSearchService.cs]: This is covered by the commit message for 45822d18ea8d1c4a1d271a3c64144747d6bebdac. I will file a follow-up issue to implement OOP caching.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: Gotcha!  Pity about this.  Tagging @jasonmalinowski for visibility.  If we do want to use JTF more in IDE, it would be good if we could get our other core primitives to play nice with it.  AsyncLazy is particularly useful and it's a pity to not be able to use it.

That said, it sounds like AsyncLazy would be wrong here anyways since this is a mutable concurrent value, not a compute-once value.  So i retract the suggestion here anyways.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 3 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: This code was not added by this pull request, but I will change this if other changes need to be made.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 5]
```diff
@@ -69,25 +72,73 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
...
+            {
+                return packageSources;
+            }
+
+            try
+            {
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: @rrelyea Do the NuGet APIs provide both of the following?

1. `IVsPackageSourceProvider.GetSources` can be called from a background thread
2. `IVsPackageSourceProvider.GetSources` is guaranteed to complete even if the UI thread is *blocked* (i.e. no main thread dependency)

If so, I can remove the use of `SwitchToMainThreadAsync` and unblock this pull request.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 6]
```diff
@@ -69,25 +72,73 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
             IThreadingContext threadingContext,
             VisualStudioWorkspaceImpl workspace,
             SVsServiceProvider serviceProvider,
-            IVsEditorAdaptersFactoryService editorAdaptersFactoryService)
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices> packageInstallerServices,
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstaller2> packageInstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageUninstaller> packageUninstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageSourceProvider> packageSourceProvider)
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: I am not sure what lazy buy us here? isn't this called right away when StartWorking get called? right now issue was GetPackageSourncesAsync being called?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 7]
```diff
@@ -106,25 +157,14 @@ protected override void EnableService()
             // Our service has been enabled.  Now load the VS package dlls.
             var componentModel = (IComponentModel)_serviceProvider.GetService(typeof(SComponentModel));
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: not used?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 6 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: by the way, isn't this whole thing already lazy through AbstractDelayStartedService (http://index/?leftProject=Microsoft.VisualStudio.LanguageServices&leftSymbol=rvtdcrtpzz7k&file=SymbolSearch%5CAbstractDelayStartedService.cs&line=21)

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 5 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: so point of this whole service is so that we can call this in code fix which Wait() on UI thread, wont UI thread switch cause dead lock here?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 8]
```diff
@@ -143,26 +183,7 @@ protected override void StartWorking()
...
-                    .Select(r => new PackageSource(r.Key, r.Value))
-                    .ToImmutableArrayOrEmpty();
-            }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
-            {
-                // These exceptions can happen when the nuget.config file is broken.
-                PackageSources = ImmutableArray<PackageSource>.Empty;
-            }
-
+            _packageSources = default;
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: whole purpose of this service is so that we can get sources from this service and save it so that code fix doesn't need to call nuget directly. 

I think if we are going to embrace JTF (I do want us to if others are okay), we need to change code fix to use JTF.Run rather than Task.Wait or Result.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 5 above)
Comment (nkolev92) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: It's safe to call from a BG thread.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 6 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: okay, understood now why this is lazy. then assuming all nuget API can be called from BG.

not sure why 
http://index/?leftProject=Microsoft.VisualStudio.LanguageServices&leftSymbol=xyflomgaagny&file=SymbolSearch%5CAbstractDelayStartedService.cs&line=53

trying to make sure it is working from UI thread. we could put push all work to BG -
http://index/?leftProject=Microsoft.VisualStudio.LanguageServices&leftSymbol=ujz4ei5xsq7x&file=Implementation%5CLanguageService%5CAbstractPackage%602.cs&line=107

Comment (heejaechang) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: feels like just moving call to Connect to BG solve all issues then?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 5 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: @nkolev92 I need an answer to the second question as well.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 6 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: @heejaechang Moving the call to background is insufficient. The call was eliminated in its entirety.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 7 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: Nice

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 8 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: > whole purpose of this service is so that we can get sources from this service and save it so that code fix doesn't need to call nuget directly.

This is fine if it occurs OOP. Currently it does not, so we opt to instead delay the load hit to the first time the feature is used.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 5 above)
Comment (nkolev92) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: The method itself doesn't need the UI thread.
It would've probably deadlocked already if it was using it.

Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: I'll update this pull request accordingly.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 8 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: Resolving this item as a duplicate of https://github.com/dotnet/roslyn/pull/35364#discussion_r279846787

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 6 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: but the other one still runs from UI thread? why not making those to run from BG as well? those 2 Connect calling from UI thread seems not necessary at all?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 3 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: Fixed

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: Commented

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 8 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: ya. since the API can call from BG, there is no issue. but since APIs can call from BG, sounds like everything can move to BG though. nothing require UI thread then.

File: src/VisualStudio/Core/Def/SymbolSearch/VisualStudioSymbolSearchService.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/SymbolSearch/VisualStudioSymbolSearchService.cs]: i don't think i quite understood this bit...  specifically, i don't quite get teh justification of:

>Package information for other sources cannot be cached until the
operation runs out-of-process.

Could you just clarify that bit?

Comment (sharwell) [reply src/VisualStudio/Core/Def/SymbolSearch/VisualStudioSymbolSearchService.cs]: The background caching results in a bunch of JIT operations and/or assembly loads even for cases where the user is not triggering this feature. RPS flagged the regression and this fixes it.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/SymbolSearch/VisualStudioSymbolSearchService.cs]: > The background caching results in a bunch of JIT operations and/or assembly loads even for cases where the user is not triggering this feature. RPS flagged the regression and this fixes it.

thanks much!

**Comment by jinujoseph** [comment]
@sharwell this will need to retargetted to 16.1 bracnh

**Comment by jinujoseph** [comment]
Keeping the fix for 16.2

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 5 above)
Comment (mavasani) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: @sharwell @nkolev92 It seems the change here indeed caused a deadlock: https://github.com/dotnet/roslyn/issues/39889. It also seems very likely to incur UI delays, invoking an external service from a code fix that blocks the user is not ideal.

Comment (nkolev92) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: I'd create an issue on NuGet/Home for this. 
NuGet has seen many delays with our synchronous services (and obviously deadlocks like now), so we are pushing for an asynchronous rewrite.

Comment (nkolev92) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: It'd be great if you can provide repro steps for the hang. 
That way we can validate any fixes. We should also look if there's any tel tracking these hangs.../cc @zivkan

Comment (mavasani) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: I have updated the description in https://github.com/dotnet/roslyn/issues/39889#issue-525249832 with the repro steps, which seems to give me pretty consistent deadlock. I can file a separate issue on NuGet/Home for the same, as we probably want to keep https://github.com/dotnet/roslyn/issues/39889 to change Roslyn code to not invoke NuGet APIs from inside a code fixer to avoid potential UI delays.
